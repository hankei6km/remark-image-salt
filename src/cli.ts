import { Readable, Writable } from 'stream'
import { unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkFrontmatter from 'remark-frontmatter'
import stringify from 'remark-stringify'
import { remarkImageSalt, RemarkImageSaltOptions } from './image-salt.js'

type Opts = {
  stdin: Readable
  stdout: Writable
  stderr: Writable
} & Required<RemarkImageSaltOptions>
const cli = async (opts: Opts): Promise<number> => {
  const { stdin, stdout, stderr, ...imageSaltOpts } = opts
  try {
    let source = ''
    await new Promise((resolve) => {
      stdin.on('data', (d) => (source = source + d))
      stdin.on('end', () => resolve(source))
    })
    const m = await unified()
      .use(remarkParse)
      .use(remarkFrontmatter, ['yaml', 'toml'])
      .use(remarkImageSalt, { ...imageSaltOpts })
      .use(stringify)
      .freeze()
      .process(source)
      .then(
        (file) => {
          return String(file)
        },
        (error) => {
          throw error
        }
      )
    stdout.write(m)
  } catch (err: any) {
    stderr.write(err.toString())
    stderr.write('\n')
    return 1
  }
  return 0
}

export default cli
