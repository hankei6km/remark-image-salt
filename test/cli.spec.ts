import { ReadStream } from 'fs'
import { PassThrough } from 'stream'
import cli from '../src/cli.js'

describe('cli()', () => {
  let stdin: PassThrough
  let stdout: PassThrough
  let stderr: PassThrough
  beforeEach(() => {
    stdin = new PassThrough()
    stdout = new PassThrough()
    stderr = new PassThrough()
  })

  it('should return stdout with exitcode=0', async () => {
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    process.nextTick(() => {
      stdin.write(
        '# test\n## test1\nimage-salt-1\n\n![image1](/path/to/iamge1.jpg)\n## test2\nimage-salt-2\n\n![image2](/path/to/iamge2.jpg)'
      )
      stdin.end()
    })

    expect(
      await cli({
        stdin,
        stdout,
        stderr,
        tagName: '',
        baseURL: '',
        keepBaseURL: false
      })
    ).toEqual(0)
    expect(outData).toMatchSnapshot()
    expect(errData).toEqual('')
  })
  it('should skip url was not matched baseURL', async () => {
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    process.nextTick(() => {
      stdin.write(
        '# test\n## test1\nimage-salt-1\n\n![image1](https://localhost:3000/path/to/iamge1.jpg)\n## test2\nimage-salt-2\n\n![image2](https://localhost:3001/path/to/iamge2.jpg)'
      )
      stdin.end()
    })

    expect(
      await cli({
        stdin,
        stdout,
        stderr,
        tagName: '',
        baseURL: 'https://localhost:3000/',
        keepBaseURL: false
      })
    ).toEqual(0)
    expect(outData).toMatchSnapshot()
    expect(errData).toEqual('')
  })
  it('should keep baseURL', async () => {
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    process.nextTick(() => {
      stdin.write(
        '# test\n## test1\nimage-salt-1\n\n![image1](https://localhost:3000/path/to/iamge1.jpg)\n## test2\nimage-salt-2\n\n![image2](https://localhost:3001/path/to/iamge2.jpg)'
      )
      stdin.end()
    })

    expect(
      await cli({
        stdin,
        stdout,
        stderr,
        tagName: '',
        baseURL: 'https://localhost:3000/',
        keepBaseURL: true
      })
    ).toEqual(0)
    expect(outData).toMatchSnapshot()
    expect(errData).toEqual('')
  })
  it('should return stderr with exitcode=1', async () => {
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    process.nextTick(() => {
      stdin.write(
        '# test\n## test1\nimage-salt-1\n\n![image1#d300x200>#](/path/to/iamge1.jpg)'
      )
      stdin.end()
    })

    expect(
      await cli({
        stdin,
        stdout,
        stderr,
        tagName: '',
        baseURL: '',
        keepBaseURL: false
      })
    ).toEqual(1)
    expect(outData).toEqual('')
    expect(errData).toMatchSnapshot()
  })
})
