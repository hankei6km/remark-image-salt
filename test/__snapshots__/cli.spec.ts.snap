// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli() should keep baseURL 1`] = `
"# test

## test1

image-salt-1

<img src=\\"https://localhost:3000/path/to/iamge1.jpg\\" alt=\\"image1\\">

## test2

image-salt-2

![image2](https://localhost:3001/path/to/iamge2.jpg)
"
`;

exports[`cli() should return stderr with exitcode=1 1`] = `
"Error: extractAttrs: invalid attrs has injected
"
`;

exports[`cli() should return stdout with exitcode=0 1`] = `
"# test

## test1

image-salt-1

<img src=\\"/path/to/iamge1.jpg\\" alt=\\"image1\\">

## test2

image-salt-2

<img src=\\"/path/to/iamge2.jpg\\" alt=\\"image2\\">
"
`;

exports[`cli() should skip url was not matched baseURL 1`] = `
"# test

## test1

image-salt-1

<img src=\\"/path/to/iamge1.jpg\\" alt=\\"image1\\">

## test2

image-salt-2

![image2](https://localhost:3001/path/to/iamge2.jpg)
"
`;
